README.md
serverless-offline-edge-lambda
=================================
A plugin for the [Serverless Framework](https://serverless.com/framework/) that simulates
the behavior of AWS CloudFront Edge Lambdas while developing offline.

### Setup

```bash
npm install --save-dev serverless
npm install --save-dev serverless-offline-edge-lambda
```

_serverless.yml_
```yaml
service:
  name: edge-lambdas

plugins:
  - serverless-offline-edge-lambda

provider:
  name: aws
  runtime: nodejs12.x

functions:
  lambda:
    handler: src/handlers.onViewerRequest
    lambdaAtEdge:
      distribution: 'WebsiteDistribution'
      eventType: 'viewer-request'
      pathPattern: '/lambda'

resources:
  Resources:
    WebsiteDistribution:
      Type: 'AWS::CloudFront::Distribution'
      Properties:
        DistributionConfig:
          DefaultCacheBehavior:
```

```bash
npx serverless offline start --port=<port>
```

#### Use with `serverless-offline`
The plugin should not be used in conjunction with `serverless-offline` because both plugins define the `offline` command.

#### Use with `serverless-plugin-cloudfront-lambda-edge`
This plugin does not handle packaging and deploying edge lambdas to the cloud. Therefore
this plugin can be used with `serverless-plugin-cloudfront-lambda-edge`. Again, doing so
is optional. The schema in _serverless.yml_ derives from that used by `serverless-plugin-cloudfront-lambda-edge`.

#### Use with Transpilers
This plugin can also be used with transpilers such as `serverless-plugin-typescript`. In the
cases where the transpiler outputs built files to a path that differs from the path
specified for the handlers (e.g. _.build/src/handers.onViewerRequest_), this plugin accepts
a configuration option `path` that it uses to resolve function handlers.

```yaml
plugins:
  - serverless-plugin-typescript

custom:
  offlineEdgeLambda:
    path: '.build'
```

For usage with `serverless-webpack` and `serverless-bundle` the config is similar but the build path changes.
```yaml
plugins:
  - serverless-webpack # or serverless-bundle

custom:
  offlineEdgeLambda:
    path: './.webpack/service/'
```

### Hot Reload Support

Hot reload for serverless-esbuild and serverless-plugin-typescript are available with extra configuration.

The watch/reload mechanism is available form serverless-webpack, but is disabled by default for esbuild and typescript.

The flag "watchReload: true" will turn on the watcher so that typescript and esbuild solutions use the watcher to hot reload the handlers.
The path to the built handlers must be specified for the watcher to work correctly.

example:
```yaml
custom:
  offlineEdgeLambda:
    path: '.esbuild/service'
    watchReload: true
```

Additional options can be used to modify the behavior of the file watcher and debounce logic (ignoreInitial, awaitWriteFinish, interval, debounce, and any other chokidar option).

example:

```yaml
custom:
  offlineEdgeLambda:
    path: '.dist/service'
    watchReload: true
    ignoreInitial: true
    awaitWriteFinish: true
    interval: 500,
    debounce: 750
```

### Options

#### `--headersFile`

Default: _undefined_

CloudFront [injects some headers](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/using-cloudfront-headers.html) into the request. You can set these by creating a JSON file and passing its path as an option.

Example:

```json
// .cf-headers.json
{
    "CloudFront-Viewer-Country": "us",
    "CloudFront-Viewer-Country-Region": "tx"
}
```

```bash
npx serverless offline start --headersFile .cf-headers.json
```


CHANGELOG.md
## [1.3.1](https://github.com/evolv-ai/serverless-offline-edge-lambda/compare/1.3.0...1.3.1) (2023-08-31)


### Bug Fixes

* **deps:** bump fs-extra and @types/fs-extra ([#637](https://github.com/evolv-ai/serverless-offline-edge-lambda/issues/637)) ([34d5765](https://github.com/evolv-ai/serverless-offline-edge-lambda/commit/34d57650f92ea282683189e5053d2b9240ba14a9))

# [1.3.0](https://github.com/evolv-ai/serverless-offline-edge-lambda/compare/1.2.3...1.3.0) (2023-08-31)


### Features

* add function name to context and event headers ([#624](https://github.com/evolv-ai/serverless-offline-edge-lambda/issues/624)) ([6be415a](https://github.com/evolv-ai/serverless-offline-edge-lambda/commit/6be415a3146f0ff13719b22f000e4eabe17b5586))

## [1.2.3](https://github.com/evolv-ai/serverless-offline-edge-lambda/compare/1.2.2...1.2.3) (2022-09-22)


### Bug Fixes

* removed toString('utf-8') in bahvior-router. ([#535](https://github.com/evolv-ai/serverless-offline-edge-lambda/issues/535)) ([c30fc10](https://github.com/evolv-ai/serverless-offline-edge-lambda/commit/c30fc1091f9b2879eea2a211ddef684ad2113fbc))

## [1.2.2](https://github.com/evolv-ai/serverless-offline-edge-lambda/compare/1.2.1...1.2.2) (2022-09-19)


### Bug Fixes

* Hot reloading with chokidar & busting of the require cache. ([#515](https://github.com/evolv-ai/serverless-offline-edge-lambda/issues/515)) ([703d1f4](https://github.com/evolv-ai/serverless-offline-edge-lambda/commit/703d1f4fe2f0cf4365425108be55d0c1cb17acff))

## [1.2.1](https://github.com/evolv-ai/serverless-offline-edge-lambda/compare/1.2.0...1.2.1) (2022-09-19)


### Bug Fixes

* fix issues with http requests ([#480](https://github.com/evolv-ai/serverless-offline-edge-lambda/issues/480)) ([85c8921](https://github.com/evolv-ai/serverless-offline-edge-lambda/commit/85c8921bd83bb039a5c4ac3ef1c8ab00349e3434))

# [1.2.0](https://github.com/evolv-ai/serverless-offline-edge-lambda/compare/1.1.2...1.2.0) (2021-10-26)


### Features

* add support for custom headers from origins ([#261](https://github.com/evolv-ai/serverless-offline-edge-lambda/issues/261)) ([3fbe995](https://github.com/evolv-ai/serverless-offline-edge-lambda/commit/3fbe9959c8a057b7b7d0d2f7361e42a211eedfd1))

## [1.1.2](https://github.com/evolv-ai/serverless-offline-edge-lambda/compare/1.1.1...1.1.2) (2021-09-01)


### Bug Fixes

* fix package build issue ([#247](https://github.com/evolv-ai/serverless-offline-edge-lambda/issues/247)) ([3923b6b](https://github.com/evolv-ai/serverless-offline-edge-lambda/commit/3923b6b8f082a18280159427bf795b6abe69aba9))

## [1.1.1](https://github.com/evolv-ai/serverless-offline-edge-lambda/compare/1.1.0...1.1.1) (2021-09-01)


### Bug Fixes

* trivial change to bump version ([#243](https://github.com/evolv-ai/serverless-offline-edge-lambda/issues/243)) ([616452f](https://github.com/evolv-ai/serverless-offline-edge-lambda/commit/616452f9b0eb09d40149bb5aafc02bcffabdd476))

# 1.0.0 (2021-08-01)


### Bug Fixes

* address issue with async iterator polyfill ([50d021f](https://github.com/evolv-ai/serverless-offline-edge-lambda/commit/50d021f030116e124f0946cf74728120916892e3))
* fix type errors ([023470f](https://github.com/evolv-ai/serverless-offline-edge-lambda/commit/023470fc40bf331e0a21f12bd7d32dff1a385564))
* make custom section optional ([e172628](https://github.com/evolv-ai/serverless-offline-edge-lambda/commit/e172628ee8042918253ae765fbac67cc474f63ba)), closes [#6](https://github.com/evolv-ai/serverless-offline-edge-lambda/issues/6)
* remove default from deprecated --cloudfrontport option ([1e127bb](https://github.com/evolv-ai/serverless-offline-edge-lambda/commit/1e127bb2f17da95d6776a872ce7f055bd1e5cefa))


### Features

* better emulation of origins and cf includeBody handling ([534b34d](https://github.com/evolv-ai/serverless-offline-edge-lambda/commit/534b34dd8cfc5c20b39f6cec9027339807528de5))
* initial commit of project files ([696021a](https://github.com/evolv-ai/serverless-offline-edge-lambda/commit/696021a4eec50c358a27743891f2fb84561cf68e))
* overhaul error handling ([d3bcf6e](https://github.com/evolv-ai/serverless-offline-edge-lambda/commit/d3bcf6eb3a274321804e39fdb0c8d20cd57814f8)), closes [#193](https://github.com/evolv-ai/serverless-offline-edge-lambda/issues/193)
* rework to run standalone from serverless-offline ([561b9dc](https://github.com/evolv-ai/serverless-offline-edge-lambda/commit/561b9dc204e1ccbca4cd652fc1e396b639473709))


src/behavior-router.ts
import { Context } from 'aws-lambda';
import bodyParser from 'body-parser';
import connect, { HandleFunction } from 'connect';
import cookieParser from 'cookie-parser';
import * as chokidar from 'chokidar';
import * as fs from 'fs-extra';
import { createServer, Server, IncomingMessage, ServerResponse } from 'http';
import { StatusCodes } from 'http-status-codes';
import * as os from 'os';
import * as path from 'path';
import { URL } from 'url';
import { debounce } from './utils/debounce';
import { HttpError, InternalServerError } from './errors/http';
import { FunctionSet } from './function-set';
import { asyncMiddleware, cloudfrontPost } from './middlewares';
import { CloudFrontLifecycle, Origin, CacheService } from './services';
import { CFDistribution, ServerlessInstance, ServerlessOptions } from './types';
import {
	buildConfig, buildContext, CloudFrontHeadersHelper, ConfigBuilder,
	convertToCloudFrontEvent, getOriginFromCfDistribution, IncomingMessageWithBodyAndCookies
} from './utils';


interface OriginMapping {
	pathPattern: string;
	target: string;
	default?: boolean;
}

export class BehaviorRouter {
	private builder: ConfigBuilder;
	private context: Context;
	private behaviors = new Map<string, FunctionSet>();
	private cfResources:  Record<string, CFDistribution>;

	private cacheDir: string;
	private fileDir: string;
	private injectedHeadersFile?: string;
	private path: string;

	private started: boolean = false;
	private origins: Map<string, Origin>;
	private restarting: boolean = false;
	private server: Server | null = null;
	private cacheService: CacheService;
	private log: (message: string) => void;

	constructor(
		private serverless: ServerlessInstance,
		private options: ServerlessOptions
	) {
		this.log = serverless.cli.log.bind(serverless.cli);

		this.builder = buildConfig(serverless);
		this.context = buildContext();

		this.cfResources = serverless.service?.resources?.Resources || {};
		this.cacheDir = path.resolve(options.cacheDir || path.join(os.tmpdir(), 'edge-lambda'));
		this.fileDir = path.resolve(options.fileDir || path.join(os.tmpdir(), 'edge-lambda'));
		this.injectedHeadersFile = options.headersFile ? path.resolve(options.headersFile) : undefined;
		this.path = this.serverless.service.custom.offlineEdgeLambda.path || '';

		fs.mkdirpSync(this.cacheDir);
		fs.mkdirpSync(this.fileDir);

		this.origins = this.configureOrigins();
		this.cacheService = new CacheService(this.cacheDir);

		if (this.serverless.service.custom.offlineEdgeLambda.watchReload) {
			this.watchFiles(path.join(this.path, '**/*'), {
				ignoreInitial: true,
				awaitWriteFinish: true,
				interval: 500,
				debounce: 750,
				...options,
			});
		}
	}

	watchFiles(pattern: any, options: any) {
		const watcher = chokidar.watch(pattern, options);
		watcher.on('all', debounce(async (eventName, srcPath) => {
			console.log('Lambda files changed, syncing...');
			await this.extractBehaviors();
			console.log('Lambda files synced');
		}, options.debounce, true));
	}

	match(req: IncomingMessage): FunctionSet | null {
		if (!req.url) {
			return null;
		}

		const url = new URL(req.url, 'http://localhost');

		for (const [, handler] of this.behaviors) {
			if (handler.regex.test(url.pathname)) {
				return handler;
			}
		}

		return this.behaviors.get('*') || null;
	}

	async start(port: number) {
		this.started = true;

		return new Promise(async (res, rej) => {
			await this.listen(port);

			// While the server is in a "restarting state" just restart the server
			while (this.restarting) {
				this.restarting = false;
				await this.listen(port, false);
			}

			res('Server shutting down ...');
		});
	}

	public hasStarted() {
		return this.started;
	}

	public isRunning() {
		return this.server !== null;
	}

	public async restart() {
		if (this.restarting) {
			return;
		}

		this.restarting = true;

		this.purgeBehaviourFunctions();
		await this.shutdown();
	}

	private async shutdown() {
		if (this.server !== null) {
			await this.server.close();
		}
		this.server = null;
	}

	private async listen(port: number, verbose: boolean = true) {
		try {
			await this.extractBehaviors();

			if (verbose) {
				this.logStorage();
				this.logBehaviors();
			}
			const app = connect();

			app.use(cloudfrontPost());
			app.use(bodyParser());
			app.use(cookieParser() as HandleFunction);
			app.use(asyncMiddleware(async (req: IncomingMessageWithBodyAndCookies, res: ServerResponse) => {
				if ((req.method || '').toUpperCase() === 'PURGE') {
					await this.purgeStorage();

					res.statusCode = StatusCodes.OK;
					res.end();
					return;
				}

				const handler = this.match(req);

				if (!handler) {
					res.statusCode = StatusCodes.NOT_FOUND;
					res.end();
					return;
				}

				const customOrigin = handler.distribution in this.cfResources ?
					getOriginFromCfDistribution(handler.pattern, this.cfResources[handler.distribution]) :
					null;

				const cfEvent = convertToCloudFrontEvent(req, this.builder('viewer-request'), this.injectedHeadersFile);

				try {
					const context = {
						...this.context,
						functionName: handler.name,
					};
					const lifecycle = new CloudFrontLifecycle(this.serverless, this.options, cfEvent,
																context, this.cacheService, handler, customOrigin);
					const response = await lifecycle.run(req.url as string);

					if (!response) {
						throw new InternalServerError('No response set after full request lifecycle');
					}

					res.statusCode = parseInt(response.status, 10);
					res.statusMessage = response.statusDescription || '';

					const helper = new CloudFrontHeadersHelper(response.headers);

					for (const { key, value } of helper.asHttpHeaders()) {
						if (value) {
							res.setHeader(key as string, value);
						}
					}

					if (response.bodyEncoding === 'base64') {
						res.end(Buffer.from(response.body ?? '', 'base64'));
					} else {
						res.end(response.body);
					}
				} catch (err) {
					this.handleError(err, res);
					return;
				}
			}));


			return new Promise(resolve => {
				this.server = createServer(app);
				this.server.listen(port);
				this.server.on('close', (e: string) => {
					resolve(e);
				});
			});
		} catch (err) {
			console.error(err);
			process.exit(1);
		}
	}

	// Format errors
	public handleError(err: HttpError, res: ServerResponse) {
		res.statusCode = err.statusCode || StatusCodes.INTERNAL_SERVER_ERROR;

		const payload = JSON.stringify(err.hasOwnProperty('getResponsePayload') ?
			err.getResponsePayload() :
			{
				code: StatusCodes.INTERNAL_SERVER_ERROR,
				message: err.stack || err.message
			}
		);

		res.end(payload);
	}

	public async purgeStorage() {
		this.cacheService.purge();
	}

	private configureOrigins(): Map<string, Origin> {
		const { custom } = this.serverless.service;
		const mappings: OriginMapping[] = custom.offlineEdgeLambda.originMap || [];

		return mappings.reduce((acc, item) => {
			acc.set(item.pathPattern, new Origin(item.target));
			return acc;
		}, new Map<string, Origin>());
	}

	private async extractBehaviors() {
		const { functions } = this.serverless.service;

		const behaviors = this.behaviors;
		const lambdaDefs = Object.entries(functions)
			.filter(([, fn]) => 'lambdaAtEdge' in fn);

		behaviors.clear();

		for await (const [, def] of lambdaDefs) {

			const pattern = def.lambdaAtEdge.pathPattern || '*';
			const distribution = def.lambdaAtEdge.distribution || '';

			if (!behaviors.has(pattern)) {
				const origin = this.origins.get(pattern);
				behaviors.set(pattern, new FunctionSet(pattern, distribution, this.log, origin, def.name));
			}

			const fnSet = behaviors.get(pattern) as FunctionSet;

			// Don't try to register distributions that come from other sources
			if (fnSet.distribution !== distribution) {
				this.log(`Warning: pattern ${pattern} has registered handlers for cf distributions ${fnSet.distribution}` +
						` and ${distribution}. There is no way to tell which distribution should be used so only ${fnSet.distribution}` +
						` has been registered.` );
				continue;
			}

			await fnSet.setHandler(def.lambdaAtEdge.eventType, path.join(this.path, def.handler));
		}

		if (!behaviors.has('*')) {
			behaviors.set('*', new FunctionSet('*', '', this.log, this.origins.get('*')));
		}
	}

	private purgeBehaviourFunctions() {
		this.behaviors.forEach((behavior) => {
			behavior.purgeLoadedFunctions();
		});
	}

	private logStorage() {
		this.log(`Cache directory: file://${this.cacheDir}`);
		this.log(`Files directory: file://${this.fileDir}`);
		console.log();
	}

	private logBehaviors() {
		this.behaviors.forEach((behavior, key) => {

			this.log(`Lambdas for path pattern ${key}` +
				(behavior.distribution === '' ? ':' : ` on ${behavior.distribution}:`)
			);

			behavior.viewerRequest && this.log(`viewer-request => ${behavior.viewerRequest.path || ''}`);
			behavior.originRequest && this.log(`origin-request => ${behavior.originRequest.path || ''}`);
			behavior.originResponse && this.log(`origin-response => ${behavior.originResponse.path || ''}`);
			behavior.viewerResponse && this.log(`viewer-response => ${behavior.viewerResponse.path || ''}`);

			console.log(); // New line
		});
	}
}
